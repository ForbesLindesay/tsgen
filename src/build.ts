import {readdirSync, readFileSync, writeFileSync} from 'fs';

function write(filename: string, str: string) {
  let old = null;
  try {
    old = readFileSync(filename, 'utf8');
  } catch (ex) {
    if (ex.code !== 'ENOENT') {
      throw ex;
    }
  }
  if (old !== str) {
    writeFileSync(filename, str);
  }
}

const builders = readdirSync(__dirname + '/walkers')
  .filter(filename => filename !== 'index.ts')
  .map(filename => filename.replace(/\.ts$/, ''));

write(
  __dirname + '/walkers/index.ts',
  `
import {writeFileSync} from 'fs';
import * as bt from 'babel-types';
import Context from '../Context';
${builders.map(n => `import walk${n} from './${n}';`).join('\n')}

export default function walk(node: bt.Statement, ctx: Context): bt.Identifier[] {
  switch (node.type) {
${builders
    .map(n => `    case '${n}': return walk${n}(node as bt.${n}, ctx);`)
    .join('\n')}
    default:
      writeFileSync(
        __dirname + '/' + node.type + '.ts',
        "import * as bt from 'babel-types';\\n" +
        "import Context from '../Context';\\n\\n" +
        "export default function walk" + node.type + "(node: bt." + node.type + ", ctx: Context): bt.Identifier[] {\\n" +
        "  throw ctx.getError('Unsupported type " + node.type + "', node);\\n" +
        "}\\n",
      );
      throw ctx.getError('Unsupported node type', node);
  }
}
`,
);

const printers = readdirSync(__dirname + '/printers')
  .filter(filename => filename !== 'index.ts')
  .map(filename => filename.replace(/\.ts$/, ''));

function special(n: string): boolean {
  return n === 'SingleImportDeclaration' || n === 'SingleVariableDeclaration';
}

write(
  __dirname + '/printers/index.ts',
  `
import {writeFileSync} from 'fs';
import * as bt from 'babel-types';
import Context, {SingleImportDeclaration, SingleVariableDeclaration} from '../Context';
${printers.map(n => `import print${n} from './${n}';`).join('\n')}

export default function print(node: bt.Node, ctx: Context): string {
  switch (node.type) {
${printers
    .map(
      n =>
        `    case '${n}': return print${n}(node as ${
          special(n) ? '' : 'bt.'
        }${n}, ctx);`,
    )
    .join('\n')}
    default:
      writeFileSync(
        __dirname + '/' + node.type + '.ts',
        "import * as bt from 'babel-types';\\n" +
        "import Context from '../Context';\\n" +
        "import print from './';\\n\\n" +
        "export default function print" + node.type + "(node: bt." + node.type + ", ctx: Context): string {\\n" +
        "  throw ctx.getError('Unsupported type " + node.type + "', node);\\n" +
        "}\\n",
      );
      throw ctx.getError('Unsupported node type', <bt.Node>node);
  }
}
`,
);
