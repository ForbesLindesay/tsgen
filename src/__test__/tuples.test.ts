import '../build';
import {lsrSync} from 'lsr';
import {convertFile} from '../';

lsrSync(__dirname + '/tuples').forEach(entry => {
  if (/\.js\.flow$/.test(entry.path)) {
    test(entry.path, () => {
      expect(convertFile(entry.fullPath)).toMatchSnapshot();
    });
  }
});
