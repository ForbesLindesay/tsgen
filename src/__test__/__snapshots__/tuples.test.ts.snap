// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`./index.js.flow 1`] = `
"declare function transitions(
  ...properties: Array<string> | [Array<string>, string]
): Object;

export default transitions;
"
`;
